File: breeze-node/src/lexer.ts
// Lexer.

import { BreezeError, Token } from "./types";

const KEYWORDS = new Set([
  "show",
  "if",
  "elif",
  "else",
  "while",
  "repeat",
  "from",
  "to",
  "for",
  "in",
  "func",
  "return",
  "break",
  "continue",
  "try",
  "catch",
  "true",
  "false",
  "none",
  "and",
  "or",
  "not",
  "import",
  "export",
]);

function isDigit(c: string): boolean {
  return c >= "0" && c <= "9";
}
function isAlpha(c: string): boolean {
  return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z") || c === "_";
}
function isAlphaNum(c: string): boolean {
  return isAlpha(c) || isDigit(c);
}

function tokenizeLine(
  src: string,
  line: number,
  tokens: Token[]
): void {
  let i = 0;
  while (i < src.length) {
    const c = src[i];

    if (c === " " || c === "\t") {
      i++;
      continue;
    }

    if (c === "#") {
      break;
    }

    // Single-quoted strings are literal (no interpolation).
    if (c === '"' || c === "'") {
      const quote = c;
      const interpolate = quote === '"';
      const parts: { text: string; expr?: string }[] = [];
      let current = "";
      let hasExpr = false;
      i++;
      while (i < src.length && src[i] !== quote) {
        if (src[i] === "\\" && i + 1 < src.length) {
          const next = src[i + 1];
          if (next === "n") current += "\n";
          else if (next === "t") current += "\t";
          else if (next === "r") current += "\r";
          else if (next === "\\") current += "\\";
          else if (next === quote) current += quote;
          else if (interpolate && next === "{") current += "{";
          else if (interpolate && next === "}") current += "}";
          else current += next;
          i += 2;
        } else if (interpolate && src[i] === "{") {
          parts.push({ text: current });
          current = "";
          hasExpr = true;
          i++;
          let depth = 1;
          let expr = "";
          while (i < src.length && depth > 0) {
            if (src[i] === "{") depth++;
            else if (src[i] === "}") {
              depth--;
              if (depth === 0) break;
            }
            expr += src[i];
            i++;
          }
          if (i >= src.length) {
            throw new BreezeError(
              "Unterminated interpolation — missing closing '}' in string",
              line
            );
          }
          i++;
          parts.push({ text: "", expr });
        } else {
          current += src[i];
          i++;
        }
      }
      if (i >= src.length) {
        throw new BreezeError("Unterminated string — missing closing quote", line);
      }
      i++;
      if (interpolate && hasExpr) {
        parts.push({ text: current });
        tokens.push({ type: "STRING", value: parts, line });
      } else if (interpolate) {
        tokens.push({ type: "STRING", value: current, line });
      } else {
        tokens.push({ type: "STRING", value: current, line });
      }
      continue;
    }

    if (isDigit(c) || (c === "." && i + 1 < src.length && isDigit(src[i + 1]))) {
      let num = "";
      let hasDot = false;
      while (i < src.length && (isDigit(src[i]) || src[i] === ".")) {
        if (src[i] === ".") {
          // Stop on second dot or when followed by a non-digit (member access).
          if (hasDot) break;
          if (i + 1 < src.length && !isDigit(src[i + 1])) break;
          hasDot = true;
        }
        num += src[i];
        i++;
      }
      const val = hasDot ? parseFloat(num) : parseInt(num, 10);
      tokens.push({ type: "NUMBER", value: val, line });
      continue;
    }

    if (isAlpha(c)) {
      let ident = "";
      while (i < src.length && isAlphaNum(src[i])) {
        ident += src[i];
        i++;
      }
      if (KEYWORDS.has(ident)) {
        tokens.push({ type: "KEYWORD", value: ident, line });
      } else {
        tokens.push({ type: "IDENT", value: ident, line });
      }
      continue;
    }

    const two = src.slice(i, i + 2);
    if (two === "==" || two === "!=" || two === "<=" || two === ">=") {
      tokens.push({ type: "OP", value: two, line });
      i += 2;
      continue;
    }

    if ("+-*/%<>=".includes(c)) {
      tokens.push({ type: "OP", value: c, line });
      i++;
      continue;
    }

    if (c === "(") {
      tokens.push({ type: "LPAREN", line });
      i++;
      continue;
    }
    if (c === ")") {
      tokens.push({ type: "RPAREN", line });
      i++;
      continue;
    }
    if (c === "[") {
      tokens.push({ type: "LBRACKET", line });
      i++;
      continue;
    }
    if (c === "]") {
      tokens.push({ type: "RBRACKET", line });
      i++;
      continue;
    }
    if (c === "{") {
      tokens.push({ type: "LBRACE", line });
      i++;
      continue;
    }
    if (c === "}") {
      tokens.push({ type: "RBRACE", line });
      i++;
      continue;
    }
    if (c === ":") {
      tokens.push({ type: "COLON", line });
      i++;
      continue;
    }
    if (c === ",") {
      tokens.push({ type: "COMMA", line });
      i++;
      continue;
    }
    if (c === ".") {
      tokens.push({ type: "DOT", line });
      i++;
      continue;
    }

    throw new BreezeError(`Unexpected character '${c}'`, line);
  }
}

// Used by the parser to re-tokenize interpolated expression segments
// like `name` in `"Hello {name}"`.
export function tokenizeExpression(src: string, line: number): Token[] {
  const tokens: Token[] = [];
  tokenizeLine(src, line, tokens);
  tokens.push({ type: "EOF", line });
  return tokens;
}

export function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  const lines = source.replace(/\t/g, "  ").split("\n");
  const indentStack: number[] = [0];

  for (let lineNum = 0; lineNum < lines.length; lineNum++) {
    const rawLine = lines[lineNum];
    const trimmed = rawLine.trim();
    const lineNo = lineNum + 1;

    // Skip blank lines and comment-only lines: they don't affect indentation.
    if (trimmed === "" || trimmed.startsWith("#")) {
      continue;
    }

    const indent = rawLine.length - rawLine.trimStart().length;
    const current = indentStack[indentStack.length - 1];

    if (indent > current) {
      indentStack.push(indent);
      tokens.push({ type: "INDENT", line: lineNo });
    } else if (indent < current) {
      while (
        indentStack.length > 1 &&
        indentStack[indentStack.length - 1] > indent
      ) {
        indentStack.pop();
        tokens.push({ type: "DEDENT", line: lineNo });
      }
      if (indentStack[indentStack.length - 1] !== indent) {
        throw new BreezeError(
          "Inconsistent indentation — does not match any previous level",
          lineNo
        );
      }
    }

    tokenizeLine(rawLine.slice(indent), lineNo, tokens);

    tokens.push({ type: "NEWLINE", line: lineNo });
  }

  while (indentStack.length > 1) {
    indentStack.pop();
    tokens.push({ type: "DEDENT", line: lines.length });
  }

  tokens.push({ type: "EOF", line: lines.length });
  return tokens;
}

File: breeze-rs/src/lexer.rs
//! Lexer.

use crate::ast::KEYWORDS;
use crate::ast::{StringPartRaw, Token, TokenKind};
use crate::error::{BreezeError, BreezeResult};

fn is_digit(c: char) -> bool {
    c.is_ascii_digit()
}
fn is_alpha(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}
fn is_alnum(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

pub fn tokenize(source: &str) -> BreezeResult<Vec<Token>> {
    let normalised = source.replace('\t', "  ");
    let lines: Vec<&str> = normalised.split('\n').collect();
    let mut tokens: Vec<Token> = Vec::new();
    let mut indent_stack: Vec<usize> = vec![0];

    for (idx, raw) in lines.iter().enumerate() {
        let line_no = idx + 1;
        let trimmed = raw.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }

        let indent = raw.len() - raw.trim_start().len();
        let current = *indent_stack.last().unwrap();

        if indent > current {
            indent_stack.push(indent);
            tokens.push(Token::new(TokenKind::Indent, line_no));
        } else if indent < current {
            while indent_stack.len() > 1 && *indent_stack.last().unwrap() > indent {
                indent_stack.pop();
                tokens.push(Token::new(TokenKind::Dedent, line_no));
            }
            if *indent_stack.last().unwrap() != indent {
                return Err(BreezeError::new(
                    "Inconsistent indentation — does not match any previous level",
                    line_no,
                ));
            }
        }

        tokenize_line(&raw[indent..], line_no, &mut tokens)?;
        tokens.push(Token::new(TokenKind::Newline, line_no));
    }

    while indent_stack.len() > 1 {
        indent_stack.pop();
        tokens.push(Token::new(TokenKind::Dedent, lines.len()));
    }

    tokens.push(Token::new(TokenKind::Eof, lines.len()));
    Ok(tokens)
}

/// Tokenize a raw expression source for the parser to re-parse inside an
/// interpolated string segment.
pub fn tokenize_expression(src: &str, line: usize) -> BreezeResult<Vec<Token>> {
    let mut tokens: Vec<Token> = Vec::new();
    tokenize_line(src, line, &mut tokens)?;
    tokens.push(Token::new(TokenKind::Eof, line));
    Ok(tokens)
}

fn tokenize_line(src: &str, line: usize, tokens: &mut Vec<Token>) -> BreezeResult<()> {
    let chars: Vec<char> = src.chars().collect();
    let mut i = 0;

    while i < chars.len() {
        let c = chars[i];

        if c == ' ' {
            i += 1;
            continue;
        }

        if c == '#' {
            break;
        }

        // Double-quoted strings allow `{expr}` interpolation; single-quoted are literal.
        if c == '"' || c == '\'' {
            let quote = c;
            let interpolate = quote == '"';
            let mut parts: Vec<StringPartRaw> = Vec::new();
            let mut current = String::new();
            let mut has_expr = false;
            i += 1;
            while i < chars.len() && chars[i] != quote {
                if chars[i] == '\\' && i + 1 < chars.len() {
                    let next = chars[i + 1];
                    match next {
                        'n' => current.push('\n'),
                        't' => current.push('\t'),
                        'r' => current.push('\r'),
                        '\\' => current.push('\\'),
                        q if q == quote => current.push(quote),
                        '{' if interpolate => current.push('{'),
                        '}' if interpolate => current.push('}'),
                        other => current.push(other),
                    }
                    i += 2;
                } else if interpolate && chars[i] == '{' {
                    // Capture raw expression source up to the matching `}`,
                    // tracking nested braces so `{f(a, b)}` and `{[1, 2]}` work.
                    parts.push(StringPartRaw::Text(std::mem::take(&mut current)));
                    has_expr = true;
                    i += 1;
                    let mut depth: i32 = 1;
                    let mut expr_src = String::new();
                    while i < chars.len() && depth > 0 {
                        if chars[i] == '{' {
                            depth += 1;
                            expr_src.push('{');
                        } else if chars[i] == '}' {
                            depth -= 1;
                            if depth == 0 {
                                break;
                            }
                            expr_src.push('}');
                        } else {
                            expr_src.push(chars[i]);
                        }
                        i += 1;
                    }
                    if i >= chars.len() {
                        return Err(BreezeError::new(
                            "Unterminated interpolation — missing closing '}' in string",
                            line,
                        ));
                    }
                    i += 1;
                    parts.push(StringPartRaw::Expr(expr_src));
                } else {
                    current.push(chars[i]);
                    i += 1;
                }
            }
            if i >= chars.len() {
                return Err(BreezeError::new(
                    "Unterminated string — missing closing quote",
                    line,
                ));
            }
            i += 1;
            if interpolate && has_expr {
                parts.push(StringPartRaw::Text(current));
                tokens.push(Token::new(TokenKind::StringParts(parts), line));
            } else {
                tokens.push(Token::new(TokenKind::String(current), line));
            }
            continue;
        }

        if is_digit(c) || (c == '.' && i + 1 < chars.len() && is_digit(chars[i + 1])) {
            let mut num = String::new();
            let mut has_dot = false;
            while i < chars.len() && (is_digit(chars[i]) || chars[i] == '.') {
                if chars[i] == '.' {
                    if has_dot {
                        break;
                    }
                    has_dot = true;
                }
                num.push(chars[i]);
                i += 1;
            }
            let val: f64 = num
                .parse()
                .map_err(|_| BreezeError::new(format!("Invalid number '{num}'"), line))?;
            tokens.push(Token::new(TokenKind::Number(val), line));
            continue;
        }

        if is_alpha(c) {
            let mut ident = String::new();
            while i < chars.len() && is_alnum(chars[i]) {
                ident.push(chars[i]);
                i += 1;
            }
            if KEYWORDS.contains(&ident.as_str()) {
                tokens.push(Token::new(TokenKind::Keyword(ident), line));
            } else {
                tokens.push(Token::new(TokenKind::Ident(ident), line));
            }
            continue;
        }

        if i + 1 < chars.len() {
            let two: String = chars[i..=i + 1].iter().collect();
            if matches!(two.as_str(), "==" | "!=" | "<=" | ">=") {
                tokens.push(Token::new(TokenKind::Op(two), line));
                i += 2;
                continue;
            }
        }

        if "+-*/%<>=".contains(c) {
            tokens.push(Token::new(TokenKind::Op(c.to_string()), line));
            i += 1;
            continue;
        }

        match c {
            '(' => tokens.push(Token::new(TokenKind::LParen, line)),
            ')' => tokens.push(Token::new(TokenKind::RParen, line)),
            '[' => tokens.push(Token::new(TokenKind::LBracket, line)),
            ']' => tokens.push(Token::new(TokenKind::RBracket, line)),
            ':' => tokens.push(Token::new(TokenKind::Colon, line)),
            ',' => tokens.push(Token::new(TokenKind::Comma, line)),
            _ => {
                return Err(BreezeError::new(
                    format!("Unexpected character '{c}'"),
                    line,
                ));
            }
        }
        i += 1;
    }

    Ok(())
}
